facia-press/app/frontpress/FapiFrontPress.scala

```diff
@@ -489,13 +453,62 @@ trait FapiFrontPress extends EmailFrontPress with Logging {
 
     contentApiResponse.map(Option(_)).fallbackTo(Future.successful(None))
   }
-
 }
 
-object CapiUrl {
-  def extractId(urlString: String): String = {
-    val uri = new URI(urlString)
-    val path = uri.getPath()
-    path.stripPrefix("/")
+object Enrichment extends Logging {
+  def enrichSnap(embedUri: Option[String], beforeEnrichment: EnrichedContent, collection: Collection, wsClient: WSClient)
+                (implicit executionContext: ExecutionContext): Future[EnrichedContent] = {
+    embedUri match {
+      case Some(embedUri) =>
+        wsClient.url(embedUri).get().map { response =>
+          Json.fromJson[EmbedJsonHtml](response.json) match {
+            case JsSuccess(embed, _) =>
+              beforeEnrichment.copy(embedHtml = Some(embed.html))
+            case _ =>
+              log.warn(s"An embed had invalid json format, and won't be pressed. $embedUri for collection ${collection.id}")
+              beforeEnrichment
+          }
+        } recover {
+          case _ => {
+            log.warn(s"A request to an embed uri failed, embed won't be pressed. $embedUri for collection ${collection.id}")
+            beforeEnrichment
+          }
+        }
+      case _ => Future(beforeEnrichment)
+    }
+  }
+
+  def enrichInteractive(atomId: Option[String], beforeEnrichment: EnrichedContent, collection: Collection, capiClient: CapiContentApiClient)
+                       (implicit executionContext: ExecutionContext): Future[EnrichedContent] = {
+    atomId match {
+      case Some(atomId) => {
+        capiClient.getResponse(ItemQuery(atomId)).map { response =>
+          response.interactive.flatMap { interactive =>
+            interactive.data match {
+              case atom: com.gu.contentatom.thrift.AtomData.Interactive =>
+                Some(beforeEnrichment.copy(embedHtml = Some(atom.interactive.html), embedCss = Some(atom.interactive.css), embedJs = atom.interactive.mainJS))
+              case _ =>
+                None
+            }
+          }.getOrElse({
+            handleInteractiveFailure(s"${atomId} for collection ${collection.id}")
+            beforeEnrichment
```

Comment:
Even though we know we're throwing an exception on the previous line, I had to add this here to keep the compiler happy (ditto the two other cases where we call `handleInteractiveFailure` below). Is there a better way to handle this? I guess leveraging types rather than throwing an exception is probably a better overall approach, which will take some more far-reaching changes.

Reply:
Looks ok :)

Reply:
@tjmw I had a go at rewriting this and came up with:

```
  def enrichInteractive(
    atomId: Option[String],
    beforeEnrichment: EnrichedContent,
    collection: Collection,
    capiClient: CapiContentApiClient)
    (implicit executionContext: ExecutionContext): Future[EnrichedContent] = {

    def asFut[A](opt: Option[A], errMsg: String): Future[A] = {
      opt match {
        case Some(thing) => Future.successful(thing)
        case None => Future.failed(new Throwable(errMsg))
      }
    }

    def enrich(response: ItemResponse): Option[EnrichedContent] = {
      for {
        interactive <- response.interactive
        enriched <- Some(interactive.data).flatMap {
          case atom: com.gu.contentatom.thrift.AtomData.Interactive =>
            Some(beforeEnrichment.copy(embedHtml = Some(atom.interactive.html), embedCss = Some(atom.interactive.css), embedJs = atom.interactive.mainJS))
          case _ => None
        }
      } yield enriched
    }

    val result = for {
      atomId <- asFut(atomId, "atomId was undefined")
      itemResponse <- capiClient.getResponse(ItemQuery(atomId))
      enriched <- asFut(enrich(itemResponse), s"unable to enrich atom $atomId")
    } yield enriched

    result.recover { case _ => {
      handleInteractiveFailure(s"${atomId} for collection ${collection.id}")
      beforeEnrichment
    }}
  }
```

Basically, it's trying to:

* coerce the types into Futures so that we can use a single `for` expression for readability and reduce the error handling repetition
* more intermediate variables again for clarity

The con is the `asFut` stuff, but it is defined close by so hopefully wouldn't confuse too much?